tsl/src/bgw_policy/policies_v2.c

```diff
@@ -0,0 +1,711 @@
+		if (!success)
+			++failures;
+	}
+	PG_RETURN_BOOL(success && (0 == failures));
+}
+
+Datum
+policies_alter(PG_FUNCTION_ARGS)
+{
+	Oid rel_oid = PG_GETARG_OID(0);
+	ContinuousAgg *cagg;
+	List *jobs;
+	ListCell *lc;
+	bool if_exists = false, found, start_found, end_found;
+	policies_info all_policies = { .refresh = NULL, .compress = NULL, .retention = NULL };
+	refresh_policy ref;
+	compression_policy comp;
+	retention_policy ret;
+
+	cagg = ts_continuous_agg_find_by_relid(rel_oid);
+	if (!cagg)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("\"%s\" is not a continuous aggregate", get_rel_name(rel_oid))));
+
+	all_policies.is_alter_policy = true;
+	all_policies.rel_oid = rel_oid;
+	all_policies.original_HT = cagg->data.raw_hypertable_id;
+	all_policies.partition_type = cagg->partition_type;
+
+	jobs = ts_bgw_job_find_by_hypertable_id(cagg->data.mat_hypertable_id);
+	if ((NIL == jobs))
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("no jobs found")));
+	foreach (lc, jobs)
+	{
+		BgwJob *job = lfirst(lc);
+		if (namestrcmp(&(job->fd.proc_name), POLICY_REFRESH_CAGG_PROC_NAME) == 0)
+		{
+			refresh_policy tmp = { .create_policy = false,
+								   .end_offset = { 0 },
+								   .end_offset_type = InvalidOid,
+								   .start_offset = { 0 },
+								   .start_offset_type = InvalidOid,
+								   .schedule_interval = job->fd.schedule_interval };
+			ref = tmp;
+			all_policies.refresh = &ref;
+			if (IS_INTEGER_TYPE(cagg->partition_type))
+			{
+				int64 start_value = ts_jsonb_get_int64_field(job->fd.config,
+															 POL_REFRESH_CONF_KEY_START_OFFSET,
+															 &start_found);
+				int64 end_value = ts_jsonb_get_int64_field(job->fd.config,
+														   POL_REFRESH_CONF_KEY_END_OFFSET,
+														   &end_found);
+				/*
+				 * If there is job then start_offset has to be there because policy is
+				 * not created without it. However if found it to be NULL, then we
+				 * want to keep it to NULL in this alter command also.
+				 */
+				all_policies.refresh->start_offset.isnull = !start_found;
+				all_policies.refresh->start_offset_type = cagg->partition_type;
+				all_policies.refresh->end_offset.isnull = !end_found;
+				all_policies.refresh->end_offset_type = cagg->partition_type;
+				switch (all_policies.refresh->start_offset_type)
+				{
+					case INT2OID:
+						all_policies.refresh->start_offset.value =
+							Int16GetDatum((int16) start_value);
+						all_policies.refresh->end_offset.value = Int16GetDatum((int16) end_value);
+						break;
+					case INT4OID:
+						all_policies.refresh->start_offset.value =
+							Int32GetDatum((int32) start_value);
+						all_policies.refresh->end_offset.value = Int32GetDatum((int32) end_value);
+						break;
+					case INT8OID:
+						all_policies.refresh->start_offset.value = Int64GetDatum(start_value);
+						all_policies.refresh->end_offset.value = Int64GetDatum(end_value);
+						break;
```

Comment:
Do we have a test for this? It seems these lines are not covered.

Reply:
Done.